spring-integration-core/src/main/java/org/springframework/integration/support/DelegatingMessageBuilder.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support;
+
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+
+/**
+ * A {@link MessageBuilder} extension, which fully delegates its operations
+ * to the provided {@link #delegate}.
+ * A custom extension can override any {@link MessageBuilder} operation on its own.
+ *
+ * @author Artem Bilan
+ * @since 4.2.9
+ */
+public class DelegatingMessageBuilder<T> extends MessageBuilder<T> {
```

Comment:
See `MessageWithThreadState` changes:
```
return new DelegatingMessageBuilder<Object>(messageBuilder) {

		@Override
		public Message<Object> build() {
			Message<Object> newMessage = super.build();
			return new MessageWithThreadState<S>(newMessage, MessageWithThreadState.this.state);
		}

};
```

I wasn't able to find any other way to retain `MessageWithThreadState` instance and provide `MessageBuild` to the caller. :cry: